who.py

```diff
@@ -8,279 +8,419 @@
+                self._session.add(db_categories_row)
+                self._session.commit()
+            # Add category and URL to indicator
+            result = (
+                self._session.query(DBIndicators)
+                .filter(DBIndicators.code == indicator_code)
+                .first()
+            )
+            if not result:
+                logger.warning(
+                    f"Indicator code {indicator_code} was not found on the "
+                    f"indicators page"
                 )
-                rows.extend(list(iterator))
-                i += indicator_limit
+                continue
+            result.url = indicator_url
+            result.category_title = category_title
+            self._session.commit()
+
+    def _create_tags(self):
+        """Use category titles to create tags"""
+        tags = []
+        replacements = {"(": "", ")": "", "/": "", ",": ""}
+        result = self._session.query(DBCategories).all()
+        for category_row in result:
+            category_title = category_row.title
+            if " and " in category_title:
+                tag_names = category_title.split(" and ")
+                for tag_name in tag_names:
+                    tags.append(
+                        multiple_replace(tag_name.strip(), replacements)
+                    )
+            else:
+                tags.append(
+                    multiple_replace(category_title.strip(), replacements)
+                )
+
+        tags = list(OrderedDict.fromkeys(tags).keys())
+        tags, _ = Vocabulary.get_mapped_tags(tags)
+
+        self._tags = tags
+
+    def _populate_indicator_data_db(self):
+        for db_row in self._session.query(DBIndicators).all():
+            indicator_name = db_row.title
+            indicator_url = db_row.url
+            indicator_code = db_row.code
+            logger.info(f"Downloading file for indicator {indicator_name}")
+            base_url = self._configuration["base_url"]
+            url = f"{base_url}api/{indicator_code}"
             try:
-                success, results = dataset.generate_resource_from_iterator(
-                    fileheaders,
-                    rows,
-                    hxltags,
-                    folder,
-                    filename,
-                    resourcedata,
-                    date_function=process_date,
-                    quickcharts=quickcharts,
+                indicator_json = self._retriever.download_json(url)
+            except (DownloadError, FileNotFoundError):
+                logger.warning(f"{url} has no data")
+                continue
+            logger.info(f"Populating DB for indicator {indicator_name}")
+            for row in indicator_json["value"]:
+                if row["SpatialDimType"] != "COUNTRY":
+                    continue
+                country_iso3 = row["SpatialDim"]
+                country_name = Country.get_country_name_from_iso3(country_iso3)
```

Comment:
For some reason after commit and add, the most time is spent in get_country_name_from_iso3 according to the profiler. Perhaps it isn't necessary to store the country name in the db if the iso3 is already in there? However, the time this takes is much less than commit or add so if they can be speeded up, it's better to focus on them

Reply:
The yappi profile has this one lower compared to cProfile so maybe not so important to look at

Reply:
That is very strange. I will try playing around with a local dictionary to see if it helps things.